bluemira/equilibria/optimisation/harmonics/toroidal_harmonics_approx_functions.py

```diff
@@ -135,6 +155,492 @@ def legendre_q(lam, mu, x, n_max=20):
+
+    ..math::
+        A_{m} = \\frac{\\mu_{0} I_{c}}{2^{5/2}} \\frac{(2m+1)!!}{2^m m!}
+        \\frac{\\sinh{\\tau_{c}}}{\\Delta_{c}^{1/2}}
+        P_{m-\\frac{1}{2}}^{-1}(\\cosh(\\tau_c))
+
+        A_{m}^{\\sin} = A_m \\sin(m \\sigma_c)
+        A_{m}^{\\cos} = A_m \\cos(m \\sigma_c)
+        A(\\tau, \\sigma) = \\sum_{m=0}^{\\infty} A_{m}^{\\cos} \\epsilon_{m} m!
+        \\sqrt{\\frac{2}{\\pi}}\\Delta^{\\frac{1}{2}} \\textbf{Q}_{m-\\frac{1}{2}}^{1}
+        (\\cosh \\tau) \\cos(m \\sigma) + A_{m}^{\\sin}\\epsilon_{m} m! \\sqrt{\\frac{2}
+        {\\pi}} \\Delta^{\\frac{1}{2}} \\textbf{Q}_{m-\\frac{1}{2}}^{1}(\\cosh \\tau)
+        \\sin(m \\sigma)
+
+
+    Parameters
+    ----------
+    eq:
+        Bluemira Equilibrium
+    th_params:
+        Dataclass holding necessary parameters for the TH approximation
+    max_degree:
+        Maximum number of degrees to calculate up to
+
+    Returns
+    -------
+    approx_coilset_psi:
+        Matrix of coilset psi values approximated using TH
+    Am_cos @ currents:
+        TH cos coefficients for required number of degrees
+    Am_sin @ currents:
+        TH sin coefficients for required number of degrees
+
+    """
+    if max_degree is None:
+        max_degree = len(th_params.th_coil_names) - 1
+
+    # Get coil positions and currents from equilibrium
+    currents = np.array([eq.coilset[name].current for name in th_params.th_coil_names])
+
+    # Initialise psi and A arrays
+    approx_coilset_psi = np.zeros_like(th_params.R)
+    A = np.zeros_like(th_params.R)
+    # Useful combination
+    Delta = np.cosh(th_params.tau) - np.cos(th_params.sigma)  # noqa: N806
+    # Get sigma values for the grid
+    sigma_mult_degree = [m * th_params.sigma for m in range(max_degree)]
+
+    epsilon = 2 * np.ones(max_degree)
+    epsilon[0] = 1
+    factorial_m = np.array([factorial(m) for m in range(max_degree)])
+    degrees = np.arange(0, max_degree)[:, None, None]
+    # TH coefficient matrix
+    Am_cos, Am_sin = coil_toroidal_harmonic_amplitude_matrix(  # noqa: N806
+        input_coils=eq.coilset,
+        th_params=th_params,
+        max_degree=max_degree,
+    )
+
+    A_coil_matrix = (  # noqa: N806
+        Am_cos[:, :, None, None]
+        * epsilon[:, None, None, None]
+        * factorial_m[:, None, None, None]
+        * np.sqrt(2 / np.pi)
+        * np.sqrt(Delta[None, None, :, :])
+        * legendre_q(degrees - 1 / 2, 1, np.cosh(th_params.tau), n_max=30)[:, None, :, :]
+        * np.cos(sigma_mult_degree)[:, None, :, :]
+        + Am_sin[:, :, None, None]
+        * epsilon[:, None, None, None]
+        * factorial_m[:, None, None, None]
+        * np.sqrt(2 / np.pi)
+        * np.sqrt(Delta[None, None, :, :])
+        * legendre_q(degrees - 1 / 2, 1, np.cosh(th_params.tau), n_max=30)[:, None, :, :]
+        * np.sin(sigma_mult_degree)[:, None, :, :]
+    )
+    A = np.array(
+        np.sum(np.einsum("ijkl, j", A_coil_matrix, currents), axis=0), dtype=float
+    )
```

Comment:
I think this is maybe slightly easier to read, definitely less calculation. I've factored out the addition and made a greater use on the einsum. I havent actually checked that they produce the same output though.

```suggestion
    Am_cos_sin = np.einsum(
        "ijkl, ikl -> ijkl", Am_cos, np.cos(sigma_mult_degrees)
    ) + np.einsum("ijkl, ikl -> ijkl", Am_sin, np.sin(sigma_mult_degrees))

    A = np.sqrt(2 / np.pi) * (
        np.einsum(
            "ijkl, i, i, kl, ikl, j",
            Am_cos_sin,
            epsilon,
            factorial_m,
            np.sqrt(Delta),
            legendre_q(degrees - 1 / 2, 1, np.cosh(th_params.tau), n_max=30),
            currents,
        )
    )
```

Reply:
Hi All, just double checked, its need to be:

"ij, ikl -> ijkl"
"ij, ikl -> ijkl"
"ijkl, i, i, kl, ikl, j -> kl" 

and will produce the same output.